Parse Attributes and attach them to their respective AST elements

dillonhicks opened the pull request:
**Issue:** #43

This PR enables parsing of attributes in various context and
attaching the attributes to their respective AST nodes. Attributes closely
follow grammar set by Rust and `ast::Attribute` elements are modeled
after `syn::Attribute`. The biggest difference is that Rune attributes
accept a wider range of literals for literal assignment, specifically :


`#[x = #{"yes": "it works"}]`


This PR **DOES NOT**:

* Provide parsing for the `ast::Attribute#input` TokenStream as a Meta
  item like syn::Attribute.
* Add the compiler support to rewrite `///` and `//!` comments as
  their respective `#[doc = ...]` & `#![doc = ...]` attributes.
* Provide compiler functionality for handling ast::Attribute elements
  in any meaningful way.  Attributes will be parsed if in a
  syntactically correct position but they will be ignored siltently.

See:

* [Rust Attribute Refrence](https://doc.rust-lang.org/reference/attributes.html)
* [syn::Attribute](https://docs.rs/syn/1.0.41/syn/struct.Attribute.html)
* *Demo*: ./scripts/attributes.rn

udoprog commented:
Thanks for the PR!

There were a couple of things I wanted to change, so I've rebased and modified the PR with the following:
* Associated attributes with `BlockExpr` and `AsyncExpr` expressions (instead of `Block`).
* Adjusted spans where attributes were added so that they include their attributes.
* Simplified parsing of attributes.
* Rejected attributes at the indexing stage (this will be replaced with the actual macro calls once attributes are actually supported beyond the parsing stage). We don't want programs with attributes to compile just yet, but soon ;).

Since you went through the work of adding `ast::Lit` (:rocket:), I've also switched everything over to use this as the representation for a literal by introducing `ast::ExprLit`.

dillonhicks commented:
> Thanks for the PR!
> 
> There were a couple of things I wanted to change, so I've rebased and modified the PR with the following:

I am glad to help! This was a bigger change than I expected it to be, so sorry about that. I would have been happy to make the changes you wanted, if even to just help familiarize myself with the code a bit more. But done is done 🕺 .

>     * Associated attributes with `BlockExpr` and `AsyncExpr` expressions (instead of `Block`).

Cool. I agree that pushing the attributes down into the block was less than ideal.

>     * Adjusted spans where attributes were added so that they include their attributes.
>
>     * Simplified parsing of attributes.

👍 

>     * Rejected attributes at the indexing stage (this will be replaced with the actual macro calls once attributes are actually supported beyond the parsing stage). We don't want programs with attributes to compile just yet, but soon ;).

The good ol' 🐔 and the 🥚  problem, I want to run a program with an unfinished feature (attributes) but that feature require the feature itself to enable the desired behavior 🤯 (I'm assuming in the future there will be a `#![feature(<feature_name>)]`)  for unstable features?). I guess I could have added a `-Oattributes` but I am not comfortable enough with the compiler yet to modify it with any confidence. I'll am going to glean some insight from looking your changes in a bit.

> 
> Since you went through the work of adding `ast::Lit` (🚀), I've also switched everything over to use this as the representation for a literal by introducing `ast::ExprLit` (to match [`syn::ExprLit`](https://docs.rs/syn/1/syn/struct.ExprLit.html)).

😎 I thought about making that change -- but this change was large enough already I felt a bit sheepish about "fixing" things that were not broken in this pass. 

On to the next task 🚀 !
